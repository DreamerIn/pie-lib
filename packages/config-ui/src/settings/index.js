import Panel from './panel';

export { Panel };

export const toggle = label => ({ type: 'toggle', label });

const toChoice = opt => {
  if (typeof opt === 'string') {
    return { label: opt, value: opt };
  } else {
    return opt;
  }
};

export const radio = function() {
  const args = Array.prototype.slice.call(arguments);
  const [label, ...opts] = args;
  return {
    type: 'radio',
    label,
    choices: opts.map(o => toChoice(o))
  };
};
