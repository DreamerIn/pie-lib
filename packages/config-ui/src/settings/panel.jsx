import PropTypes from 'prop-types';
import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import debug from 'debug';
import Toggle from './toggle';
import TwoChoice from '../two-choice';
const log = debug('pie-lib:config-ui:settings:panel');
import _ from 'lodash';

const labelValue = {
  label: PropTypes.string,
  value: PropTypes.string
};

const baseTypes = {
  label: PropTypes.string,
  value: PropTypes.string,
  onChange: PropTypes.func
};

const Radio = props => {
  return (
    <TwoChoice
      value={props.value}
      header={props.label}
      one={props.choices[0]}
      two={props.choices[1]}
      onChange={props.onChange}
    />
  );
};

Radio.propTypes = { ...baseTypes, choices: PropTypes.arrayOf(PropTypes.shape(labelValue)) };

const ToggleWrapper = props => (
  <Toggle label={props.label} checked={!!props.value} toggle={props.onChange} />
);

ToggleWrapper.propTypes = {
  ...baseTypes,
  value: PropTypes.bool
};

const tagMap = {
  toggle: ToggleWrapper,
  radio: Radio
};

const Group = withStyles(theme => ({
  group: {
    marginTop: theme.spacing.unit * 3
  },
  groupHeader: {
    fontSize: '10px',
    fontWeight: 500
  }
}))(props => {
  const { classes, model, label, group } = props;

  return (
    <div className={classes.group}>
      <div className={classes.groupHeader}>{label.toUpperCase()}</div>

      {Object.keys(group).map(key => {
        if (!group[key]) {
          return null;
        }
        const tagProps = { ...group[key], key, value: _.get(model, key) };
        const Tag = tagMap[tagProps.type];
        return <Tag key={key} {...tagProps} onChange={v => props.onChange(key, v)} />;
      })}
    </div>
  );
});

export class Panel extends React.Component {
  static propTypes = {
    model: PropTypes.object,
    groups: PropTypes.object,
    onChange: PropTypes.func
  };

  changeModel = (key, value) => {
    log('[changeModel]', key, value);
    const { onChange } = this.props;
    const model = { ...this.props.model };

    _.set(model, key, value);

    onChange(model, key);
  };

  render() {
    const { groups, model } = this.props;
    log('render:', model);
    return (
      <div>
        {Object.keys(groups).map(g => (
          <Group label={g} key={g} model={model} group={groups[g]} onChange={this.changeModel} />
        ))}
      </div>
    );
  }
}

export default Panel;
