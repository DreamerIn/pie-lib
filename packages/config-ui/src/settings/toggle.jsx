import React from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import { withStyles } from '@material-ui/core/styles';
import Switch from '@material-ui/core/Switch';

const Toggle = withStyles(theme => ({
  toggle: {
    display: 'flex',
    width: '100%',
    justifyContent: 'space-between'
  },
  label: {
    paddingTop: theme.spacing.unit * 2
  }
}))(({ checked, label, toggle, classes }) => (
  <div className={classes.toggle}>
    <Typography className={classes.label}>{label}</Typography>
    <Switch checked={checked} onChange={e => toggle(e.target.checked)} />
  </div>
));

Toggle.propTypes = {
  checked: PropTypes.bool,
  label: PropTypes.string.isRequired,
  toggle: PropTypes.func.isRequired
};

export default Toggle;
