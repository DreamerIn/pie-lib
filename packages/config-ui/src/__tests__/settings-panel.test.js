import React from 'react';
import { shallow } from 'enzyme';
import { Panel } from '../settings/panel';
import { toggle, radio } from '../settings';

describe('Settings Panel', () => {
  let w;
  let onChange = jest.fn();
  let configure = {
    orientationLabel: 'Orientation',

    settingsOrientation: true,
    editChoiceLabel: false
  };
  let model = {
    choiceAreaLayout: 'vertical',
    configure
  };

  let groups = ({ configure }) => ({
    'Group 1': {
      choiceAreaLayout: configure.settingsOrientation && {
        type: 'radio',
        label: configure.orientationLabel,
        choices: [{ label: 'opt1', value: 'opt1' }, { label: 'opt2', value: 'opt2' }]
      },
      'configure.editChoiceLabel': { type: 'toggle', label: 'Edit choice label' }
    }
  });

  const wrapper = extras => {
    return shallow(<Panel model={model} onChange={onChange} groups={groups(model)} {...extras} />);
  };

  describe('snapshot', () => {
    it('renders', () => {
      w = wrapper();

      expect(w).toMatchSnapshot();
    });

    it('does not render radio buttons', () => {
      w = wrapper({
        groups: groups({
          configure: {
            ...configure,
            settingsOrientation: false
          }
        })
      });

      expect(w).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    describe('onChange gets called', () => {
      it('updates root props', () => {
        w.instance().changeModel('test', false);

        expect(onChange).toBeCalledWith(
          {
            ...model,
            test: false
          },
          'test'
        );
      });

      it('updates nested props', () => {
        w.instance().changeModel('configure.test.test', true);

        expect(onChange).toBeCalledWith(
          {
            ...model,
            configure: {
              ...configure,
              test: {
                test: true
              }
            }
          },
          'configure.test.test'
        );
      });
    });
  });
});

describe('toggle', () => {
  it('returns a toggle type object', () => {
    const setting = toggle('Label');

    expect(setting).toEqual({
      label: 'Label',
      type: 'toggle'
    });
  });
});

describe('radio', () => {
  it('returns a radio type object', () => {
    const setting = radio('Radio', 'one', 'two');

    expect(setting).toEqual({
      label: 'Radio',
      type: 'radio',
      choices: [
        {
          label: 'one',
          value: 'one'
        },
        {
          label: 'two',
          value: 'two'
        }
      ]
    });
  });
});
