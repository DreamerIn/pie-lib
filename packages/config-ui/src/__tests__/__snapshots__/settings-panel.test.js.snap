// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings Panel snapshot does not render radio buttons 1`] = `
<div>
  <WithStyles(Component)
    group={
      Object {
        "choiceAreaLayout": false,
        "configure.editChoiceLabel": Object {
          "label": "Edit choice label",
          "type": "toggle",
        },
      }
    }
    key="Group 1"
    label="Group 1"
    model={
      Object {
        "choiceAreaLayout": "vertical",
        "configure": Object {
          "editChoiceLabel": false,
          "orientationLabel": "Orientation",
          "settingsOrientation": true,
        },
      }
    }
    onChange={[Function]}
  />
</div>
`;

exports[`Settings Panel snapshot renders 1`] = `
<div>
  <WithStyles(Component)
    group={
      Object {
        "choiceAreaLayout": Object {
          "choices": Array [
            Object {
              "label": "opt1",
              "value": "opt1",
            },
            Object {
              "label": "opt2",
              "value": "opt2",
            },
          ],
          "label": "Orientation",
          "type": "radio",
        },
        "configure.editChoiceLabel": Object {
          "label": "Edit choice label",
          "type": "toggle",
        },
      }
    }
    key="Group 1"
    label="Group 1"
    model={
      Object {
        "choiceAreaLayout": "vertical",
        "configure": Object {
          "editChoiceLabel": false,
          "orientationLabel": "Orientation",
          "settingsOrientation": true,
        },
      }
    }
    onChange={[Function]}
  />
</div>
`;
